marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java

```diff
@@ -44,27 +46,82 @@ private OutputCallerImpl getCaller() {
 
     @Override
     public InputStream[] call() {
-        return call(null, null, null);
+        return call(newCallContext());
     }
+
     @Override
+    @Deprecated
     public InputStream[] call(InputStream endpointState, SessionState session, InputStream workUnit) {
-        checkAllowedArgs(endpointState, session, workUnit);
-        return getCaller().arrayCall(getClient(), endpointState, session, workUnit);
+        CallContext callContext = newCallContext().withEndpointState(endpointState).withSessionState(session)
+                .withWorkUnit(workUnit);
+        return call(callContext);
     }
 
     @Override
+    @Deprecated
     public OutputEndpoint.BulkOutputCaller bulkCaller() {
-        return new BulkOutputCallerImpl(this);
+        return bulkCaller(newCallContext());
+    }
+
+    @Override
+    public InputStream[] call(CallContext callContext) {
```

Comment:
The challenge is that 

- the deprecated methods must return the endpoint state as the first item in the array for backward compatibility.
- the new methods must return the array without the endpoint state because, with call context, that hack isn't needed.

The implementation in the patch:

- breaks backward compatibility for the deprecated methods because the endpoint state isn't returned
- adds inefficiency to the new methods because they always create a new array

Ultimately, we might want to push the decision down into the array caller.
 
For now, maybe the expedient approach is as follows:

1. move the current code of InputStream[] call(CallCont) into a private InputStream[] call(CallCont, withEndpointState boolean) method
2.  if withEndpointState is true, return the values as is; otherwise, return the copy that drops the endpoint state
3.  call the private method with true from the deprecated call() methods and with false from the new method
4.  create a Jira task for considering refactoring to push awareness of the CallContext down into the array caller

The same issue applies to the output for the input-output call implementation.